### PR TITLE
Unit test TTLReportReconciler controller

### PR DESCRIPTION
**Pull request by padlar**
Before making changes to the TTLReportReconciler, unit test the functionality

## Related issues
- related #237

### CODE REVIEW

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+		reportUpdateTimestamp time.Duration
+		ttlAnnotationMissing  bool
+		expectError           bool
+		reportDeleted         bool
+		ttlStr                string
+		invalidReportName     bool
+	}{
+		{
+			name:                  "Report timestamp < TTL",
+			reportUpdateTimestamp: -15 * 60 * time.Minute, // < 24h TTL
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:57]:
```suggestion
			reportUpdateTimestamp: -15 * time.Hour, // < 24h TTL
```

And also the others. Easier to read.

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+	// logger object
+	logger := log.Log.WithName("testing")
+
+	// create TTLReport controller
+	instance := vulnerabilityreport.TTLReportReconciler{Logger: logger, Config: config, Clock: clock}
+
+	tests := []struct {
+		name                  string
+		reportUpdateTimestamp time.Duration
+		ttlAnnotationMissing  bool
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:49]: I think you can get rid of this by treating the empty `ttlStr` as this case.

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+			ttlAnnotationMissing: true,
+			expectError:          false, // Ignoring report without TTL set
+		},
+		{
+			name:        "invalid TTL in the annotation",
+			ttlStr:      "badtime",
+			expectError: true,
+		},
+		{
+			name:              "invalid report name",
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:80]: I think this is more like non-existing report. And I don't think you need an extra property for it. Can't you just use the name?

**Review by erikgb — CHANGES_REQUESTED** [review]

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 1 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: good catch! fixed it

File: pkg/vulnerabilityreport/ttl_report.go [hunk 4]
```diff
@@ -2,8 +2,8 @@ package vulnerabilityreport
 
 import (
 	"context"
-	"fmt"
 	"time"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report.go]: yes fixed it

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 2 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: this flag removes the entire annotation when set, so helpful in testing missing entire annotation and invalid duration value

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 3 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: yes fixed the name. extra property is required to set the name to an empty string.

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 2 above)
Comment (erikgb) [reply pkg/vulnerabilityreport/ttl_report_test.go]: I think you can use the same conditional with `ttlStr != ""`.

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	// logger object
+	logger := log.Log.WithName("testing")
+
+	// create TTLReport controller
+	instance := vulnerabilityreport.TTLReportReconciler{Logger: logger, Config: config, Clock: clock}
+
+	tests := []struct {
+		name                  string
+		reportUpdateTimestamp time.Duration
+		expectError           bool
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:49]:
```suggestion
		wantErr           bool
```

Nit

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+		reportDeleted         bool
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:50]:
```suggestion
		wantReportDeleted         bool
```

Nit

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		ttlStr                string
+		invalidReportName     bool
+		ttlAnnotationMissing  bool
+	}{
+		{
+			name:                  "Report timestamp < TTL",
+			reportUpdateTimestamp: -15 * time.Hour, // < 24h TTL
+			expectError:           false,
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:58]: False is the zero value of bool, so I would prefer if you don't set it in those cases. To make the test more compact an easier to read. Applies to all similar cases.

**Review by erikgb — CHANGES_REQUESTED** [review]

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 2 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: removed

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 5 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: fixed

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 6 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: updated

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 3 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: updated the code to explicitly set the name

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		wantError             bool
+		wantReportDeleted     bool
+		ttlStr                string
+		invalidReportName     bool
+	}{
+		{
+			name:                  "Report timestamp < TTL",
+			reportUpdateTimestamp: -15 * time.Hour, // < 24h TTL
+			wantError:             false,
+			wantReportDeleted:     false,
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:63]:
```suggestion
```

In this case I do not think this adds any value.

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+			ttlStr:                "24h",
+		},
+		{
+			name:                  "Report timestamp exceeds TTL",
+			reportUpdateTimestamp: -25 * time.Hour, // > 24 TTL
+			wantError:             false,
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:69]:
```suggestion
```

Adds no value

File: pkg/vulnerabilityreport/ttl_report_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+		{
+			name:      "invalid TTL in the annotation",
+			ttlStr:    "badtime",
+			wantError: true,
+		},
+		{
+			name:              "non-existent report name",
+			invalidReportName: true,  // sets the report name to empty string
+			wantError:         false, // missing/invalid report ignored
+			wantReportDeleted: false,
```
Comment (erikgb) [inline pkg/vulnerabilityreport/ttl_report_test.go:87]:
```suggestion
```

No meaning in expecting a delete of something that doesn't exist.

**Review by erikgb — CHANGES_REQUESTED** [review]
Almost LGTM, but I am suggesting to remove some zero-values from the test fixtures. But keeping the cases where we want to express the intent.

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 9 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: removed

File: pkg/vulnerabilityreport/ttl_report_test.go (hunk 10 above)
Comment (padlar) [reply pkg/vulnerabilityreport/ttl_report_test.go]: removed

**Review by josedonizetti — APPROVED** [review]
Thank you!

**Pull request merged**